### ISSUE TITLE
Supporting custom errors

### ISSUE THREAD

**Issue by amitmahbubani** [twitchtv/twirp#239]
Loosely related to #238 

Twirp is opinionated with respect to the error generation and serialization, with it's own:
1. Error interface
2. Error codes
3. Error code to status code mapping and,
4. Error serialization interface.

... and this is fairly sufficient for internal services. However, for a Twirp service exposed externally to REST consumers, I'd like to be able to customise my errors. For example, 

An error message whose corresponding JSON encoding will be exposed to clients:
```protobuf
message Error {
    // Error type, for classification
    string type = 1;

    // public error code
    string code = 2;

    // description of the error, in text
    string description = 3;

    // Request field that caused the error
    string field = 4;

    // Reserved for future cases
    reserved 5 to 10;

    // Other details/metadata
    repeated google.protobuf.Any details = 11;
} 
```

Would Twirp consider support for custom error type definitions and serialization extensions?

**Comment by dpolansky** [comment]
Twirp errors do not seem like the right fit for the use case you describe of presenting a message with a specific shape to end users. Even in the case of internal service-to-service communication, adding an additional field that contains error data to an RPC's response message may be practical than including such data in an error (depending on the use case).

Unfortunately, I don't think Twirp will support custom error definitions (beyond what is already available) or mechanisms to customize the serialization of errors.

It seems more practical to implement a layer in front of your Twirp service that prepares Twirp responses for presentation to end users than it would be to modify Twirp to support that use case. Have you considered that approach? Some folks at Twitch have implemented such solutions and may be able to respond to specific concerns you may have.

**Comment by amitmahbubani** [comment]
> Even in the case of internal service-to-service communication, adding an additional field that contains error data to an RPC's response message may be practical than including such data in an error (depending on the use case).

Agree with this, and I would have taken this route *if* I was able to control the HTTP status code on the response. Inspecting the response payload for an "error" message at the client end, when the status_code=200, would be bad practice.

> It seems more practical to implement a layer in front of your Twirp service that prepares Twirp responses for presentation to end users than it would be to modify Twirp to support that use case. Have you considered that approach?

Yes. For some quick context, we (@razorpay) have been using Twirp for internal services pretty well. However, widespread internal adoption is not possible since we have a public REST API. Reg other options:
1. Transform before Twirp, within a service: either fork Twirp, or with a custom `ResponseWriter` -- both have maintainability issues.
2. Transform at an API gateway (@kong in our case) - requires us to rely on Twirp error's limited details. For example, with Meta being `map[string]string`, trying to send out a native float field would require hacky inspection. 

These are do-able of course, but it comes down to us patching and adding layers of indirection to workaround Twirp limits, going against why we chose Twirp in the first place: the out-of-the-box simplicity over gRPC. IMO, extensibility does not have to be a trade-off over simplicity -- would have been great if we could have extended on Twirp when needed.

Would be interesting to know how others workaround this and #238.
